File: csharp/Link.Foundation.Links.Notation/StreamParser.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Link.Foundation.Links.Notation
{
    /// <summary>Incrementally emits complete top-level Links Notation records.</summary>
    public sealed class StreamParser
    {
        private const int DefaultMaxBufferSize = 10 * 1024 * 1024;
        private const string BeforeReference = " \t\n\r(:";
        private const string BeforeComment = " \t\n\r";

        private readonly Parser _parser;
        private readonly StringBuilder _buffer = new();
        private readonly StringBuilder _currentLine = new();
        private readonly List<Link<string>> _links = new();
        private int? _baseIndentation;
        private bool _lineClassified;
        private int _offset;
        private int _line;
        private int _column;
        private int _segmentOffset;
        private int _segmentLine;
        private bool _ended;
        private int _maxBufferSize = DefaultMaxBufferSize;

        /// <summary>Creates a stream backed by a default canonical parser.</summary>
        public StreamParser() : this(new Parser()) { }

        /// <summary>Creates a stream backed by <paramref name="parser"/>.</summary>
        public StreamParser(Parser parser)
        {
            _parser = parser ?? throw new ArgumentNullException(nameof(parser));
            Reset();
        }

        /// <summary>Raised once for every completed link.</summary>
        public event Action<Link<string>>? LinkParsed;

        /// <summary>Whether emitted links are retained for <see cref="Finish()"/> and <see cref="Drain"/>.</summary>
        public bool Collect { get; set; } = true;

        /// <summary>The largest unresolved record accepted by the stream.</summary>
        public int MaxBufferSize
        {
            get => _maxBufferSize;
            set => _maxBufferSize = value > 0
                ? value
                : throw new ArgumentOutOfRangeException(nameof(value), "Maximum buffer size must be positive.");
        }

        /// <summary>The point immediately after the last character written.</summary>
        public StreamPosition Position => new(
            _offset,
            _line,
            _column,
            _buffer.Length + _currentLine.Length);

        /// <summary>Consumes a chunk and returns links made complete by it.</summary>
        public IReadOnlyList<Link<string>> Write(string chunk)
        {
            ArgumentNullException.ThrowIfNull(chunk);
            if (_ended) throw new InvalidOperationException("Cannot write after Finish().");

            var emitted = new List<Link<string>>();
            foreach (var character in chunk)
            {
                _currentLine.Append(character);
                _offset++;

                if (character == '\n')
                {
                    _buffer.Append(_currentLine);
                    _currentLine.Clear();
                    _lineClassified = false;
                    _line++;
                    _column = 1;
                }
                else
                {
                    if (!_lineClassified && character != ' ' && character != '\t' && character != '\r')
                    {
                        _lineClassified = true;
                        if (!(_parser.Comments && character == Comments.Comment))
                        {
                            StartContentLine(LeadingSpaces(_currentLine), emitted);
                        }
                    }
                    _column++;
                }

                if (_buffer.Length + _currentLine.Length > MaxBufferSize)
                {
                    throw new ArgumentException(
                        $"Buffered record exceeds maximum size of {MaxBufferSize} characters.",
                        nameof(chunk));
                }
            }
            return emitted;
        }

        /// <summary>Finishes the stream and returns all undrained links.</summary>
        public IReadOnlyList<Link<string>> Finish() => Finish(string.Empty);

        /// <summary>Consumes a final chunk and finishes the stream.</summary>
        public IReadOnlyList<Link<string>> Finish(string chunk)
        {
            if (_ended)
            {
                if (chunk.Length > 0) throw new InvalidOperationException("Cannot write after Finish().");
                return Collect ? _links.ToArray() : Array.Empty<Link<string>>();
            }

            var emitted = chunk.Length == 0
                ? new List<Link<string>>()
                : new List<Link<string>>(Write(chunk));
            var document = _buffer.ToString() + _currentLine;
            if (document.Length > 0)
            {
                IList<Link<string>> parsed;
                try
                {
                    parsed = _parser.Parse(document);
                }
                catch (ParseException error)
                {
                    throw new StreamParseException(
                        error,
                        _segmentOffset + error.Offset,
                        _segmentLine + error.Line - 1,
                        error.Column);
                }
                Publish(parsed, emitted);
                AdvanceSegment(document);
            }

            _buffer.Clear();
            _currentLine.Clear();
            _baseIndentation = null;
            _ended = true;
            return Collect ? _links.ToArray() : emitted;
        }

        /// <summary>Returns and forgets links retained since the previous call.</summary>
        public IReadOnlyList<Link<string>> Drain()
        {
            var links = _links.ToArray();
            _links.Clear();
            return links;
        }

        /// <summary>Reuses the parser while preserving its options and event handlers.</summary>
        public StreamParser Reset()
        {
            _buffer.Clear();
            _currentLine.Clear();
            _baseIndentation = null;
            _lineClassified = false;
            _links.Clear();
            _offset = 0;
            _line = 1;
            _column = 1;
            _segmentOffset = 0;
            _segmentLine = 1;
            _ended = false;
            return this;
        }

        /// <summary>Lazily parses a sequence of chunks as a native enumerable.</summary>
        public static IEnumerable<Link<string>> ParseChunks(
            IEnumerable<string> chunks,
            Parser? parser = null)
        {
            var stream = new StreamParser(parser ?? new Parser()) { Collect = false };
            foreach (var chunk in chunks)
            {
                foreach (var link in stream.Write(chunk)) yield return link;
            }
            foreach (var link in stream.Finish()) yield return link;
        }

        /// <summary>Lazily parses an asynchronous sequence of chunks.</summary>
        public static async IAsyncEnumerable<Link<string>> ParseChunksAsync(
            IAsyncEnumerable<string> chunks,
            Parser? parser = null,
            [EnumeratorCancellation] CancellationToken cancellationToken = default)
        {
            var stream = new StreamParser(parser ?? new Parser()) { Collect = false };
            await foreach (var chunk in chunks.WithCancellation(cancellationToken))
            {
                foreach (var link in stream.Write(chunk)) yield return link;
            }
            foreach (var link in stream.Finish()) yield return link;
        }

        private void StartContentLine(int indentation, List<Link<string>> emitted)
        {
            if (_buffer.Length > 0 &&
                _baseIndentation.HasValue &&
                indentation <= _baseIndentation.Value &&
                StructurallyComplete(_buffer.ToString(), _parser.Comments))
            {
                try
                {
                    var document = _buffer.ToString();
                    var parsed = _parser.Parse(document);
                    Publish(parsed, emitted);
                    AdvanceSegment(document);
                    _buffer.Clear();
                    _baseIndentation = null;
                }
                catch (ParseException)
                {
                    // A rejected prefix can become valid when more input arrives.
                }
            }

            _baseIndentation ??= indentation;
        }

        private void Publish(IEnumerable<Link<string>> links, List<Link<string>> emitted)
        {
            foreach (var link in links)
            {
                emitted.Add(link);
                if (Collect) _links.Add(link);
                LinkParsed?.Invoke(link);
            }
        }

        private void AdvanceSegment(string document)
        {
            _segmentOffset += document.Length;
            foreach (var character in document)
            {
                if (character == '\n') _segmentLine++;
            }
        }

        private static int LeadingSpaces(StringBuilder line)
        {
            var indentation = 0;
            while (indentation < line.Length && line[indentation] == ' ') indentation++;
            return indentation;
        }

        private static bool StructurallyComplete(string document, bool comments)
        {
            var depth = 0;
            for (var position = 0; position < document.Length; position++)
            {
                var character = document[position];
                if (Quotes.IsQuote(character) && Follows(document, position, BeforeReference))
                {
                    var end = Quotes.ReferenceEnd(document, position);
                    if (end < 0) return false;
                    position = end - 1;
                    continue;
                }
                if (comments && character == Comments.Comment && Follows(document, position, BeforeComment))
                {
                    var newline = document.IndexOf('\n', position);
                    if (newline < 0) break;
                    position = newline;
                    continue;
                }
                if (character == '(') depth++;
                else if (character == ')') depth--;
            }
            return depth == 0;
        }

        private static bool Follows(string document, int position, string allowed) =>
            position == 0 || allowed.IndexOf(document[position - 1]) >= 0;
    }

    /// <summary>An absolute stream position and the unresolved record size.</summary>
    public readonly record struct StreamPosition(int Offset, int Line, int Column, int Buffered);
}

File: java/src/main/java/io/github/linkfoundation/linksnotation/StreamParser.java
package io.github.linkfoundation.linksnotation;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/** Incrementally emits complete top-level Links Notation records. */
public final class StreamParser {
  private static final int DEFAULT_MAX_BUFFER_SIZE = 10 * 1024 * 1024;
  private static final String BEFORE_REFERENCE = " \t\n\r(:";
  private static final String BEFORE_COMMENT = " \t\n\r";

  /** Point immediately after the last character written. */
  public record Position(int offset, int line, int column, int buffered) {}

  private final Parser parser;
  private Consumer<Link> onLink;
  private boolean collect = true;
  private int maxBufferSize;
  private final StringBuilder buffer = new StringBuilder();
  private final StringBuilder currentLine = new StringBuilder();
  private Integer baseIndentation;
  private boolean lineClassified;
  private final List<Link> links = new ArrayList<>();
  private int offset;
  private int line;
  private int column;
  private int segmentOffset;
  private int segmentLine;
  private boolean ended;

  /** Creates a stream parser with default canonical parser options. */
  public StreamParser() {
    this(new Parser());
  }

  /** Creates a stream parser backed by the supplied canonical parser. */
  public StreamParser(Parser parser) {
    if (parser == null) {
      throw new IllegalArgumentException("Parser must not be null");
    }
    this.parser = parser;
    this.maxBufferSize = Math.min(DEFAULT_MAX_BUFFER_SIZE, parser.getMaxInputSize());
    reset();
  }

  /** Register a consumer called once for every completed link. */
  public StreamParser onLink(Consumer<Link> consumer) {
    this.onLink = consumer;
    return this;
  }

  /** Enable or disable retaining emitted links for finish and drain. */
  public StreamParser collect(boolean value) {
    this.collect = value;
    return this;
  }

  /** Set the largest unresolved record accepted by this stream. */
  public StreamParser maxBufferSize(int value) {
    if (value < 1) {
      throw new IllegalArgumentException("Maximum buffer size must be positive");
    }
    this.maxBufferSize = value;
    return this;
  }

  /** Consume a chunk and return the links made complete by it. */
  public List<Link> write(String chunk) throws ParseException {
    if (chunk == null) {
      throw new IllegalArgumentException("Input must not be null");
    }
    if (ended) {
      throw new IllegalStateException("Cannot write after finish()");
    }

    List<Link> emitted = new ArrayList<>();
    for (int index = 0; index < chunk.length(); index++) {
      char character = chunk.charAt(index);
      currentLine.append(character);
      offset++;

      if (character == '\n') {
        buffer.append(currentLine);
        currentLine.setLength(0);
        lineClassified = false;
        line++;
        column = 1;
      } else {
        if (!lineClassified && character != ' ' && character != '\t' && character != '\r') {
          lineClassified = true;
          if (!(parser.isCommentsEnabled() && character == Comments.COMMENT)) {
            startContentLine(leadingSpaces(currentLine), emitted);
          }
        }
        column++;
      }

      if (buffer.length() + currentLine.length() > maxBufferSize) {
        throw new IllegalArgumentException(
            "Buffered record exceeds maximum size of " + maxBufferSize + " characters");
      }
    }
    return emitted;
  }

  /** Finish the stream and return all undrained links. */
  public List<Link> finish() throws ParseException {
    return finish("");
  }

  /** Consume a final chunk, finish the stream, and return its result. */
  public List<Link> finish(String chunk) throws ParseException {
    if (ended) {
      if (!chunk.isEmpty()) {
        throw new IllegalStateException("Cannot write after finish()");
      }
      return collect ? List.copyOf(links) : List.of();
    }

    List<Link> emitted = chunk.isEmpty() ? new ArrayList<>() : write(chunk);
    String document = buffer.toString() + currentLine;
    if (!document.isEmpty()) {
      List<Link> parsed;
      try {
        parsed = parser.parse(document);
      } catch (ParseException error) {
        throw new StreamParseException(error, segmentOffset, segmentLine, 1);
      }
      publish(parsed, emitted);
      advanceSegment(document);
    }

    buffer.setLength(0);
    currentLine.setLength(0);
    baseIndentation = null;
    ended = true;
    return collect ? List.copyOf(links) : emitted;
  }

  /** Return and forget links retained since the previous drain. */
  public List<Link> drain() {
    List<Link> drained = List.copyOf(links);
    links.clear();
    return drained;
  }

  /** Reuse the parser while preserving configuration and its consumer. */
  public StreamParser reset() {
    buffer.setLength(0);
    currentLine.setLength(0);
    baseIndentation = null;
    lineClassified = false;
    links.clear();
    offset = 0;
    line = 1;
    column = 1;
    segmentOffset = 0;
    segmentLine = 1;
    ended = false;
    return this;
  }

  /** Return the absolute stream position and unresolved buffer size. */
  public Position position() {
    return new Position(offset, line, column, buffer.length() + currentLine.length());
  }

  /** Lazily parse an iterable of chunks as a native Java Stream. */
  public static Stream<Link> parseChunks(Iterable<String> chunks) {
    Iterator<String> input = chunks.iterator();
    StreamParser parser = new StreamParser().collect(false);
    Deque<Link> ready = new ArrayDeque<>();
    Spliterator<Link> iterator =
        new Spliterators.AbstractSpliterator<>(
            Long.MAX_VALUE, Spliterator.ORDERED | Spliterator.NONNULL) {
          private boolean finished;

          @Override
          public boolean tryAdvance(Consumer<? super Link> action) {
            while (ready.isEmpty() && !finished) {
              try {
                if (input.hasNext()) {
                  ready.addAll(parser.write(input.next()));
                } else {
                  ready.addAll(parser.finish());
                  finished = true;
                }
              } catch (ParseException error) {
                throw new StreamParseRuntimeException(error);
              }
            }
            Link next = ready.pollFirst();
            if (next == null) {
              return false;
            }
            action.accept(next);
            return true;
          }
        };
    return StreamSupport.stream(iterator, false);
  }

  private void startContentLine(int indentation, List<Link> emitted) {
    if (buffer.length() > 0
        && baseIndentation != null
        && indentation <= baseIndentation
        && structurallyComplete(buffer, parser.isCommentsEnabled())) {
      try {
        List<Link> parsed = parser.parse(buffer.toString());
        publish(parsed, emitted);
        advanceSegment(buffer.toString());
        buffer.setLength(0);
        baseIndentation = null;
      } catch (ParseException ignored) {
        // A prefix rejected by the canonical parser may become valid with more input.
      }
    }
    if (baseIndentation == null) {
      baseIndentation = indentation;
    }
  }

  private void publish(List<Link> parsed, List<Link> emitted) {
    for (Link link : parsed) {
      emitted.add(link);
      if (collect) {
        links.add(link);
      }
      if (onLink != null) {
        onLink.accept(link);
      }
    }
  }

  private void advanceSegment(String document) {
    segmentOffset += document.length();
    segmentLine += (int) document.chars().filter(character -> character == '\n').count();
  }

  private static int leadingSpaces(CharSequence line) {
    int indentation = 0;
    while (indentation < line.length() && line.charAt(indentation) == ' ') {
      indentation++;
    }
    return indentation;
  }

  private static boolean structurallyComplete(CharSequence document, boolean comments) {
    String text = document.toString();
    int depth = 0;
    for (int position = 0; position < text.length(); position++) {
      char character = text.charAt(position);
      if (isQuote(character) && follows(text, position, BEFORE_REFERENCE)) {
        int end = Parser.quotedReferenceEnd(text, position);
        if (end < 0) {
          return false;
        }
        position = end - 1;
        continue;
      }
      if (comments && character == Comments.COMMENT && follows(text, position, BEFORE_COMMENT)) {
        int newline = text.indexOf('\n', position);
        if (newline < 0) {
          break;
        }
        position = newline;
        continue;
      }
      if (character == '(') {
        depth++;
      } else if (character == ')') {
        depth--;
      }
    }
    return depth == 0;
  }

  private static boolean isQuote(char character) {
    return character == '"' || character == '\'' || character == '`';
  }

  private static boolean follows(String document, int position, String allowed) {
    return position == 0 || allowed.indexOf(document.charAt(position - 1)) >= 0;
  }
}
